grails-app/domain/org/pih/warehouse/order/OrderAdjustment.groovy

```diff
@@ -56,4 +57,8 @@ class OrderAdjustment implements Serializable {
     def getTotalAdjustments() {
         return amount ?: percentage ? orderItem ? orderItem?.subtotal * (percentage/100) : order.subtotal * (percentage/100) : 0
     }
+
+    def getIsInvoiced() {
+        return InvoiceItem.list().find { it.orderAdjustment == this && it.invoice.dateSubmitted } ? true : false
```

Comment:
Write criteria query (InvoiceItem.createCriteria()) or HQL (InvoiceItem.executeQuery()) here as this will become a bottleneck with enough invoice items in the system. In fact, this should just be a query that returns true or false.

Reply:
I wanted to do it but orderAdjustment and orderItem are derived fields on Invoice Item. Can createCriteria or executeQuery handle this?

Reply:
At the very least you should do 
```
Boolean getIsInvoiced() { 
  return InvoiceItem.createCriteria().list { 
      invoice {
          notNull("dateSubmitted")
      }
  }.find { it.orderAdjustment == this }
}
```

You might not be able to reference the derived property in the criteria query but you should be able to achieve this using a join in HQL. Start by writing SQL to answer this question for a single order adjustment and then we can help you convert that into a HQL or criteria query. 

In general, we should always write SQL if something won't work with our current associations.

Reply:
```
select * 
from invoice_item 
join invoice on invoice_item.invoice_id = invoice.id 
join order_adjustment_invoice on order_adjustment_invoice.invoice_item_id = invoice_item.id 
where invoice.date_submitted is not null;
```

Reply:
You should probably add a join parameter for the order adjustment and then pass the order adjustment ID as a parameter.

Reply:
If possible turn this into an HQL query. If necessary we could add an association.